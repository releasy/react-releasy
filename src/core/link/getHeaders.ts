import { UploadableMap } from 'relay-runtime';

const getHeaders = (uploadables?: UploadableMap): Object => {
  if (uploadables) {
    return {
      Accept: '*/*',
    };
  }

  return {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };
};

export default getHeaders;
