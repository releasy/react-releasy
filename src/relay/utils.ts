import { CacheConfig, RequestNode } from 'relay-runtime';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);
