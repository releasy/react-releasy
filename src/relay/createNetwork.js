// @flow
import { Network } from 'relay-runtime';
import RelayNetworkLogger from 'relay-runtime/lib/RelayNetworkLogger';

import type { CacheConfig, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import { isMutation, mustForceFetch } from './utils';

import type { Config } from '../index';

const createNetwork = (config: Config): Network => {
  const fetchFn = config.networkLogger
    ? RelayNetworkLogger.wrapFetch(config.link.fetch)
    : config.link.fetch;

  return Network.create(
    async (
      request: RequestNode,
      variables: Variables,
      cacheConfig: CacheConfig,
      uploadables?: UploadableMap,
    ) => {
      const queryID = request.text;

      // no cache configured, just fetch the data from server
      if (!config.cache) {
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // on mutations, clear cache and then fetch the data
      if (isMutation(request)) {
        config.cache.clear();
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // verify if data exists on cache
      const fromCache = config.cache.get(queryID, variables);
      if (fromCache !== null && !mustForceFetch(cacheConfig)) {
        return fromCache;
      }

      const fromServer = await fetchFn(request, variables, cacheConfig, uploadables);
      if (fromServer) {
        config.cache.set(queryID, variables, fromServer);
      }

      return fromServer;
    }
  );
};

export default createNetwork;
