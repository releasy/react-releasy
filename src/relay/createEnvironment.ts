import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
