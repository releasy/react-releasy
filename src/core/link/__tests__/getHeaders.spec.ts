import { createUploadables } from '../../../../tests/fixtures/relay';

import getHeaders from '../getHeaders';

it('should create default headers', () => {
  const headers = getHeaders();

  expect(headers).toMatchSnapshot();
});

it('should create headers for uploadables', () => {
  const uploadables = createUploadables();

  const headers = getHeaders(uploadables);

  expect(headers).toMatchSnapshot();
});
