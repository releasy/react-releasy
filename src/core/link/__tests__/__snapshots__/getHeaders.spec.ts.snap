// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create default headers 1`] = `
Object {
  "Accept": "application/json",
  "Content-Type": "application/json",
}
`;

exports[`should create headers for uploadables 1`] = `
Object {
  "Accept": "*/*",
}
`;
